celery/backends/cosmosdbsql.py

```diff
@@ -0,0 +1,222 @@
+# -*- coding: utf-8 -*-
+"""The CosmosDB/SQL backend for Celery."""
+from __future__ import absolute_import, unicode_literals
+
+import re
+
+from kombu.utils import cached_property
+from kombu.utils.encoding import bytes_to_str
+
+from celery.exceptions import ImproperlyConfigured
+from celery.utils.log import get_logger
+
+from .base import KeyValueStoreBackend
+
+try:
+    import pydocumentdb
+    from pydocumentdb.document_client import DocumentClient
+    from pydocumentdb.documents import ConnectionPolicy
+    from pydocumentdb.documents import ConsistencyLevel
+    from pydocumentdb.documents import PartitionKind
+    from pydocumentdb.errors import HTTPFailure
+    from pydocumentdb.retry_options import RetryOptions
+except ImportError:  # pragma: no cover
+    pydocumentdb = DocumentClient = ConsistencyLevel = PartitionKind = \
+        HTTPFailure = ConnectionPolicy = RetryOptions = None  # noqa
+
+__all__ = ("CosmosDBSQLBackend",)
+
+
+ERROR_NOT_FOUND = 404
+ERROR_EXISTS = 409
+
+URL_RE = re.compile(r"cosmosdbsql://"
+                    r"AccountEndpoint=(?P<endpoint>[^;]+);"
+                    r"AccountKey=(?P<key>[^;]+);")
+
+LOGGER = get_logger(__name__)
+
+
+class CosmosDBSQLBackend(KeyValueStoreBackend):
+    """CosmosDB/SQL backend for Celery."""
+
+    def __init__(self,
+                 url=None,
+                 database_name=None,
+                 collection_name=None,
+                 consistency_level=None,
+                 max_retry_attempts=None,
+                 max_retry_wait_time=None,
+                 *args,
+                 **kwargs):
+        super(CosmosDBSQLBackend, self).__init__(*args, **kwargs)
+
+        if pydocumentdb is None:
+            raise ImproperlyConfigured(
+                "You need to install the pydocumentdb library to use the "
+                "CosmosDB backend.")
+
+        conf = self.app.conf
+
+        self._endpoint, self._key = self._parse_url(url)
```

Comment:
@c-w I am guessing the key is used to authenticate and should be kept secret. Since the URL does not follow the standard URL authentication format (`user:password@host`), it may be exposed during logging. Have you perhaps checked this in the worker logs especially?

Reply:
See https://github.com/celery/celery/pull/5185/commits/2e45ac4f03bee7a3f2ddcfd61a3eff8579dcdd34